### PR TITLE
35 wko physics modelling

### PR DESCRIPTION
**Pull request by KobusVanRooyen**
- Cyclic stored heat for the WKO added
- Some formatting
- Kept current energy loss = f(heat_loss_coef) on the warm well. Cold well heat gain excluded for now

### CODE REVIEW

File: tests/models/wko/src/example.py [hunk 1]
```diff
@@ -167,8 +170,16 @@ def path_goals(self):
...
+        try:
+            self.energy_system_components["heat_source"]
+
+            for s in self.energy_system_components["heat_source"]:
+                goals.append(MinimizeSourcesHeatGoal(s))
+        except KeyError:
+            ...
+        except Exception:
+            traceback.print_exc()
+            sys.exit(1)
```
Comment (FJanssen-TNO) [inline tests/models/wko/src/example.py:182]: what is the reason for this try except statement? maybe add a comment.

File: tests/test_cold_demand.py [hunk 2]
```diff
@@ -64,11 +67,102 @@ def test_airco(self):
...
+                #     discharged. -> WKO in heating mode
+                # if heat balance over the year is required, then this needs to change ?????? TBC
+                # for ates_id in self.energy_system_components.get("low_temperature_ates", []):
+                #     stored_volume = self.state_vector(f"{ates_id}.Stored_volume")
+                #     volume_usage = 0.0
+                #     for itstep in range(len(self.times()) - 1):
+                #         volume_usage = (
+                #             volume_usage + stored_volume[itstep + 1] - stored_volume[itstep]
+                #         )
+                #     constraints.append((volume_usage, 0.0, 0.0))
```
Comment (FJanssen-TNO) [inline tests/test_cold_demand.py:137]: instead of looping over the timesteps, you could consider to set the value of the volume at step 0 and the last step equal. Then it uses the end volume as an initial state at the start of the year.

**Review by FJanssen-TNO — COMMENTED** [review]
I haven't checked the test case yet as I saw some TODOs from your side. Some first comments are below.

File: tests/test_cold_demand.py (hunk 2 above)
Comment (KobusVanRooyen) [reply tests/test_cold_demand.py:142]: I agree

File: tests/models/wko/src/example.py (hunk 1 above)
Comment (KobusVanRooyen) [reply tests/models/wko/src/example.py:182]: Thanks for noting this. I think I added this due to my network not having a heat_source (heat pump) in starting testing. It only had an ATES with a cold and hot demand. I realized afterwards that this does not really make sense, because then the heat and cold demand must be in balance which is not realistic. I have deleted this "try....

File: tests/test_cold_demand.py (hunk 2 above)
Comment (KobusVanRooyen) [reply tests/test_cold_demand.py]: Code update. Currently it is commented out because one has to choose between either heat balance or volume balance over the time period

**Comment by KobusVanRooyen** [comment]
@FJanssen-TNO  ready for review

**Review by FJanssen-TNO — COMMENTED** [review]
I have resolved the issue when the heatlosses are turned off. The bug was caused by the temperature selection at the pipes to determine the heat, when heatlosses are included the heat should be limited by max(Temperature, T_ground) while for pipes below ground temperature and when no heatlosses occur, the used temperature should be Temperature.

File: tests/models/wko/src/example.py [hunk 3]
```diff
@@ -166,6 +166,7 @@ def path_goals(self):
         The appended list of goals
         """
         goals = super().path_goals().copy()
+        self.energy_system_components["heat_source"]
```
[the remark this answers is not in the document]
Comment (KobusVanRooyen) [reply tests/models/wko/src/example.py:169]: Agreed. Line of code removed

**Comment by KobusVanRooyen** [comment]
@FJanssen-TNO ready for review again. Please note that I updated the test case -> check that heat gain, heat loss and no heat loss actually happen. I specifically did not put in hard-coded heat gain/loss values for the comparison, but rather checking if the values are + and - as expected. But now I wonder if I should not have put in hard-coded values. What do you think?

**Review by FJanssen-TNO — APPROVED** [review]
@KobusVanRooyen the PR is approved, everything is solved.
Lets add a test where we check the pipe heatloss value in a separate PR.
